Modifying parts of optimization to comply with new config file format

BhargavaKrishnaSreepathi opened the pull request:
This PR does two things:

1. Creates optimization_constants.py file to incorporate all the constants used in the optimization folder.  The corresponding dependencies have been modified

2. Modifies files in optimization folder to comply with the new config file.

It can be tested by running 'demand_main' followed by 'optimization_main' files for a reference case. I have tested on both reference-case-open and reference-case-zug

JIMENOFONSECA commented on cea/globalvar.py:
```diff
@@ -266,18 +120,17 @@ def __init__(self):
 
-        # Circulating Pump
-        self.etaPump = 0.8
+        self.ELEC_PRICE = 0.2 * self.EURO_TO_CHF / 1000.0  # default 0.2
```
I believe all this variables related to prices should either be part of the cost database or the config file. 
you can add one variable called conversion_factor. which basically converts all the values from your database to whatever currency you want.

JIMENOFONSECA commented on cea/technologies/cogeneration.py:
```diff
@@ -141,14 +141,14 @@ def CC_Op(wdot_W, gt_size_W, fuel, tDH_K, gV) :
-    (eta, mdot_kgpers, texh_K, mdotfuel_kgpers) = GT_partLoadParam(wdot_W, gt_size_W, eta0, mdot0_kgpers, fuel, gV)
-    (qdot_W, wdotfin_W) = ST_Op(mdot_kgpers, texh_K, tDH_K, fuel, gV)
+    (eta0, mdot0_kgpers) = GT_fullLoadParam(gt_size_W, fuel, gV, optimization_constants)
+    (eta, mdot_kgpers, texh_K, mdotfuel_kgpers) = GT_partLoadParam(wdot_W, gt_size_W, eta0, mdot0_kgpers, fuel, gV, optimization_constants)
```
I see the variable gV still being passed one. should not it have been completely remove with this PR?

JIMENOFONSECA commented on cea/technologies/cogeneration.py:
```diff
@@ -175,29 +175,29 @@ def GT_fullLoadParam(gt_size_W, fuel, gV):
     return eta0, mdot0_kgpers
 
-def GT_partLoadParam(wdot_W, gt_size_W, eta0, mdot0, fuel, gV):
+def GT_partLoadParam(wdot_W, gt_size_W, eta0, mdot0, fuel, gV, optimization_constants):
```
the same, I think gV should not be passed anymroe...

JIMENOFONSECA commented on cea/technologies/substation.py:
```diff
@@ -19,7 +19,7 @@
 # Substation model
 
-def substation_main(locator, total_demand, building_names, gv, Flag):
+def substation_main(locator, total_demand, building_names, gv, optimization_constants, Flag):
```
The same here, gv is still being passed on.

BhargavaKrishnaSreepathi commented:
Updated the PR to address the comments. Also, added a new file `prices.py` in `cea\optimization`, which imports the prices from the database as a `class`, and is further being used in multiple files. As per my understanding, this prevents repetitive importing of data from the excel files. 
This PR also tries to decrease the dependency on the `globalvariables` file significantly, and I can see that we can get rid of `globalvariables` file after one more `localizing` session.

@JIMENOFONSECA  and @daren-thomas, do take a look at the PR and be as critical as possible.

daren-thomas commented on cea/optimization/preprocessing/decentralized_buildings.py:
```diff
@@ -95,32 +95,32 @@ def calc_new_load(mdot, TsupDH, Tret, gv):
             # Bloom box emissions within the FC: 773 lbs / MWh_el (and 1 lbs = 0.45 kg)
             # http://www.carbonlighthouse.com/2011/09/16/bloom-box/
-            result[2][6] += 1.51 * Qgas * 3600E-6 - gv.EL_TO_OIL_EQ * Qelec * 3600E-6 # MJ-oil-eq
+            result[2][6] += 1.51 * Qgas * 3600E-6 - optimization_constants.EL_TO_OIL_EQ * Qelec * 3600E-6 # MJ-oil-eq
```
what does `3600E-6` mean? another magic number that could be replaced by a constant at the top of the file... the kind of constant that would make reading the line easier to understand. is it a conversion factor?

daren-thomas commented on cea/optimization/preprocessing/preprocessing_main.py:
```diff
@@ -129,23 +131,19 @@ def __init__(self, locator, building_names):
-    locator = cea.inputlocator.InputLocator(scenario_path=config.scenario)
+    locator = cea.inputlocator.InputLocator(scenario=config.scenario)
     total_demand = pd.read_csv(locator.get_total_demand())
     building_names = pd.read_csv(locator.get_total_demand())['Name']
```
this happens a lot in CEA code and @martin-mosteiro  and I just found out that we need a unified way of getting the names of the buildings in a scenario. There's an open issue and you are gladly invited to come up with a nice solution to share :-)

daren-thomas replied:
(see issue #934)

daren-thomas commented on cea/optimization/prices.py:
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+"""
+Data required for Slave from Master
+This File sets all variables for the slave optimization, that have to be set by the Master
```
this comment doesn't say anything. the file is called `prices`. maybe talk about that instead?

daren-thomas commented on cea/optimization/prices.py:
```diff
@@ -0,0 +1,32 @@
+class Prices(object):
+    def __init__(self, locator, config):
+        pricing = pd.read_excel(locator.get_supply_systems(config.region), sheetname="Pricing")
+        self.NG_PRICE = pricing[pricing['Description'] == 'ng_price'].iloc[0]['value']
```
use lower-case for variable names. see pep8.

daren-thomas commented on cea/optimization/prices.py:
```diff
@@ -0,0 +1,32 @@
+        self.NG_PRICE_zernez = pricing[pricing['Description'] == 'ng_price_zernez'].iloc[0]['value']
+        self.BG_PRICE_zernez = pricing[pricing['Description'] == 'bg_price_zernez'].iloc[0]['value']
+        self.ELEC_PRICE = pricing[pricing['Description'] == 'elec_price'].iloc[0]['value']
+        self.cPump = pricing[pricing['Description'] == 'cpump'].iloc[0]['value']
```
cPump? what does that stand for? this whole method would be a brilliant place to document what these variables mean.

daren-thomas commented on cea/optimization/slave/least_cost.py:
```diff
@@ -839,55 +839,55 @@ def calc_primary_energy_and_CO2(Q_source_data_W, Q_coldsource_data_W, E_PP_el_da
-    Eprim_from_Sewage = np.sum(Q_HPSew_gen_W) / COP_HPSew_avg  * gv.SEWAGEHP_TO_OIL_STD * gv.Wh_to_J / 1.0E6
-    Eprim_from_GHP    = np.sum(Q_GHP_gen_W) / COP_GHP_avg * gv.GHP_TO_OIL_STD  * gv.Wh_to_J / 1.0E6
-    Eprim_from_HPLake = np.sum(Q_HPLake_gen_W) / COP_HPLake_avg * gv.LAKEHP_TO_OIL_STD  * gv.Wh_to_J / 1.0E6
+    Eprim_from_Sewage = np.sum(Q_HPSew_gen_W) / COP_HPSew_avg  * optimization_constants.SEWAGEHP_TO_OIL_STD * gv.Wh_to_J / 1.0E6
```
here again, `/ 1.0E6` - could be just a constant. idn... `* Wh_toMJ`?

daren-thomas replied:
oops. should be `WH_TO_MJ`

daren-thomas commented on cea/optimization/slave/seasonal_storage/design_operation.py:
```diff
@@ -17,7 +17,7 @@
 
 
 def Storage_Design(CSV_NAME, SOLCOL_TYPE, T_storage_old_K, Q_in_storage_old_W, locator,
```
variable names should be lowercase. function parameters are variables.

daren-thomas commented on cea/technologies/boilers.py:
```diff
@@ -104,18 +104,18 @@ def cond_boiler_op_cost(Q_therm_W, Q_design_W, T_return_to_boiler_K, BoilerFuelT
 
 
     if BoilerFuelType == 'BG':
```
'BG' should probably be written as a constant, like, `BIO_GAS`? (i'm assuming that's what it stands for?)

And since you went to the trouble of creating a class for prices, this would be an *excellent* place to refactor the `if` to a method: `gas_price = prices.gas_price(boiler_fuel_type)`

daren-thomas replied:
(apply that to the stuff below too)

daren-thomas commented on cea/technologies/solar/photovoltaic_thermal.py:
```diff
@@ -547,15 +547,37 @@ def calc_PVT_module(tilt_angle_deg, IAM_b_vector, IAM_d, I_direct_vector, I_diff
 # investment and maintenance costs
 
-def calc_Cinv_PVT(P_peak, gv):
+def calc_Cinv_PVT(PVT_peak_kW, locator, config, technology = 0):
```
what is a `0` technology? where would you look that up? there has to be a more communicative way to specify this...

daren-thomas commented on cea/technologies/thermal_network/thermal_network_matrix.py:
```diff
@@ -1615,14 +1621,14 @@ def run_as_script(scenario_path=None):
     set_diameter = True  # this does a rule of max and min flow to set a diameter. if false it takes the input diameters
 
-    path, list_network, files = os.walk(locator.get_input_network_folder(network_type[1])).next()
+    path, list_network, files = os.walk(locator.get_input_network_folder(network_type[0])).next()
```
yeah... so what exactly is a `network_type[0]`?

daren-thomas commented on cea/inputlocator.py:
```diff
@@ -152,7 +153,9 @@ def get_optimization_network_layout_pipes_file(self):
         Optimized network layout files for pipes of district heating networks
         """
-        return os.path.join(self.get_optimization_network_layout_folder(), "DH_AllEdges.csv")
+        import cea.config
```
can we make `config` a parameter instead of creating it locally? Otherwise, this creates a subtle bug for runs that use a modified config file (e.g. command-line parameters or the arcgis interface)

daren-thomas commented on cea/optimization/master/cost_model.py:
```diff
@@ -14,7 +14,7 @@
 import numpy as np
 import pandas as pd
-
+from cea.optimization.optimization_constants import *
```
we could actually just call the module `constants`, since it's full name is `cea.optimization.constants`

daren-thomas commented on cea/optimization/master/evaluation.py:
```diff
@@ -215,81 +221,81 @@ def calc_master_to_slave_variables(individual, Qmax, locator, gv, optimization_c
     #CHP units with NG & furnace with biomass wet
     if individual[0] == 1 or individual[0] == 3:
-        if optimization_constants.Furnace_allowed == 1:
+        if Furnace_allowed == 1:
```
I would really like this to be a boolean... unless... what are valid values for this variable?

daren-thomas commented on cea/optimization/master/evaluation.py:
```diff
@@ -306,15 +312,15 @@ def calc_master_to_slave_variables(individual, Qmax, locator, gv, optimization_c
 
 
-def checkNtw(individual, ntwList, locator, gv, optimization_constants):
+def checkNtw(individual, ntwList, locator, gv):
```
and when we do get around to some refactoring, this could be called `check_network` - abreviations should be used sparingly...

daren-thomas commented on cea/optimization/optimization_constants.py:
```diff
@@ -234,6 +136,102 @@
 Furn_min_Load = 0.2  # Minimum load possible (does not affect Model itself!)
 Furn_min_electric = 0.3  # Minimum load for electricity generation in furnace plant
 
+ZernezFlag = 0
```
I think these should all be booleans?

daren-thomas commented on cea/technologies/thermal_network/thermal_network_matrix.py:
```diff
@@ -1600,37 +1603,29 @@ def read_properties_from_buildings(building_names, buildings_demands, property):
     gv = cea.globalvar.GlobalVariables()
-
-    import cea.config
     config = cea.config.Configuration()
```
this is a bug. you already have a variable called `config`. you shouldn't be creating your own instances of `cea.config.Configuration()` anyway...

BhargavaKrishnaSreepathi commented:
@daren-thomas @JIMENOFONSECA : i think most of the comments were answered in the subsequent commits. Do let me know if you have further comments
